src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs

```diff
@@ -0,0 +1,1151 @@
+        }
+
+        [Fact]
+        public void RecursiveCalling_RefReadOnlyExtensionMethods()
+        {
+            var code = @"
+public struct S1
+{
+    public int i;
+
+    public void Mutate()
+    {
+        System.Console.Write(i--);
+    }
+}
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this S1 obj)
+    {
+        if (obj.i > 0)
+        {
+            obj.Mutate();
+            obj.PrintValue();
+        }
+    }
+}
+public class Program
+{
+    public static void Main()
+    {
+        var obj = new S1 { i = 5 };
+        obj.PrintValue();
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "54321");
+        }
+
+        [Fact]
+        public void AmbiguousRefnessForExtensionMethods_RValue()
+        {
+            var code = @"
+public static class Ext1
+{
+    public static void Print(ref int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Ext2
+{
+    public static void Print(ref readonly int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public class Program
+{
+    public static void Main()
+    {
+        0.Print();                  // Error
+        Ext2.Print(0);              // Ok
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (20,11): error CS1061: 'int' does not contain a definition for 'Print' and no extension method 'Print' accepting a first argument of type 'int' could be found (are you missing a using directive or an assembly reference?)
+                //         0.Print();                  // Error
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "Print").WithArguments("int", "Print").WithLocation(20, 11));
+        }
+
+        [Fact]
+        public void AmbiguousRefnessForExtensionMethods_LValue()
+        {
+            var code = @"
+public static class Ext1
+{
+    public static void Print(ref int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Ext2
+{
+    public static void Print(ref readonly int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public class Program
+{
+    public static void Main()
+    {
+        int value = 0;
+        value.Print();              // Error
+        
+        Ext1.Print(ref value);      // Ok
+        Ext2.Print(value);          // Ok
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (21,15): error CS1061: 'int' does not contain a definition for 'Print' and no extension method 'Print' accepting a first argument of type 'int' could be found (are you missing a using directive or an assembly reference?)
+                //         value.Print();              // Error
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "Print").WithArguments("int", "Print").WithLocation(21, 15));
```

Comment:
Yes. An authoring mistake on my end. Thanks for pointing this out.

---
In reply to: [122029236](https://github.com/dotnet/roslyn/pull/19703#discussion_r122029236) [](ancestors = 122029236)